### PR TITLE
Refactor ec2_vpc_dhcp_option to use boto3 (breaking change)

### PR DESCRIPTION
**Pull request by jillr**
##### SUMMARY
* Convert to boto3
* More consistent use of check_mode
* Use AwsRetry
* Catch BotoCoreError, ClientError

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/modules/ec2_vpc_dhcp_option.py

### CODE REVIEW

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 1]
```diff
@@ -190,91 +190,256 @@
...
+        #     if existing_config[existing_index]['Value'] != new_config[new_index]['Value']:
+        #         new_config[new_index]['Value'] = existing_config[existing_index]['Value']
+        if existing_config and not new_index:
+            new_config.append(existing_config[existing_index])
+            changed = True
+
+    return changed, new_config
+
+
+def normalize_config(option_config):
```
Comment (goneri) [inline plugins/modules/ec2_vpc_dhcp_option.py:389]: This seems to be a good candidate for an unit-test.

Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:389]: I'm still working through all these data transformations (I keep finding cases where they don't quite work) but you're absolutely right, they should all be covered.

**Comment by jillr** [comment]
@tremble What would you think about the boto3 version of this module having occasionally slightly different return data, but substantially more consistent (and documented) returns?  Things like, in the boto2 module we sometimes returned all DhcpConfigurations keys, we sometimes only returned the ones that were set, and sometimes (such as the `  - name: add and remove tags` test) returned no keys.  I'm thinking about always returning all keys, with `None` set if there's no value, in the interest of simplifying the internal logic.

Whether something is returned as a string or a list should also be more predictable now, though I haven't modified the integration tests to confirm for certain yet.
(All the places where we currently have things like ` dhcp_options.new_options['domain-name'] in ["{{ aws_domain_name }}", ["{{ aws_domain_name }}"]]` right now.)

Basically this has turned into a full rewrite, so I'm considering making it part of a 2.0 of the collection and being a bit freer with making the code long-term easier to maintain over a perfect 1:1 matching of the boto2 module.

**Comment by tremble** [comment]
I'm generally good with returning keys where we didn't before.

Changing what existing return values return (strings vs lists) makes me less comfortable and might be better handled with a breaking release.  It's really hard to communicate this to users and 'deprecate' return values.

In general I'd like to be very deliberate about performing a 2.x release since it becomes the point at which we'll probably get asked to start backporting fixes between branches.

- Are there other places where we'd like to 'break' return values for consistency?
  - Do we have anywhere other than ec2_vpc_igw_info still returning tags as a list of dicts rather than a simple list?
  - Do we want to remove **all** the old boto2 compatibility return hacks?
  - De we want to push towards more consistent return values, maybe adding the new return types in 2.x and removing the old ones in a 3.x
- Would a 2.x release be a good time to nuke all the _facts aliases?
- Would a 2.x release be a good time to make module names a little more consistent? (aws_ is a little redundant and we have a lot of ec2_ modules people might not realise are tied to the EC2 API)

**Comment by jillr** [comment]
A 2.0 would definitely be more deliberate, I'd see that as happening at the same time we move the first batch of modules out of community.aws and 2.0 that collection (hopefully later this spring).  That would leave this PR WIP until that time, but free me up to simplify the logic quite a bit.

We put the facts aliases deprecation notices as 2021-12-01, so I think we're best off waiting until after then regardless of collection version.

I would be in favor of starting to update any return changes and module names that we can in 2.x.

**Comment by jillr** [comment]
This is mostly done, minus unit tests.  There are some lingering bugs noted in the integration test file but I'm inclined to land this already extensive rewrite and handle those later as they're are for features or functions that the module has never provided anyway.

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 2]
```diff
@@ -190,91 +220,252 @@
...
-    start_time = time()
-    while time() < start_time + 300:
+
+def fetch_dhcp_options_for_vpc(client, module, vpc_id):
+    try:
+        vpcs = client.describe_vpcs(aws_retry=True, VpcIds=[vpc_id])['Vpcs']
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Unable to describe vpc {0}".format(vpc_id))
+
+    if len(vpcs) != 1 or vpcs[0]['IsDefault']:
```
Comment (tremble) [inline plugins/modules/ec2_vpc_dhcp_option.py:277]: Why return `None` if it's the Default VPC?

File: tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml [hunk 3]
```diff
@@ -261,17 +257,18 @@
...
+  - debug:
+      msg: '{{ dhcp_options.new_options.keys() }}'
+
   - assert:
       that:
         - dhcp_options.changed
         - dhcp_options.new_options
         # FIXME extra keys are returned unpredictably
-        - dhcp_options.new_options.keys() | list | sort is superset(['netbios-name-servers', 'netbios-node-type', 'ntp-servers'])
+        - dhcp_options.new_options.keys() | list | sort == ['netbios-name-servers', 'netbios-node-type', 'ntp-servers']
```
Comment (tremble) [inline tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml:268]:
```suggestion
        - dhcp_options.new_options.keys() | list | sort == ['netbios-name-servers', 'netbios-node-type', 'ntp-servers']
```

I was thinking that maybe failing if a new option was added is the best option here.  However, I think if someone wants to add additional info, we should probably get them to return the boto3 style data alongside the original boto2 data.

Maybe worth a comment here about why we fail if *extra* return values are passed?

**Review by tremble — CHANGES_REQUESTED** [review]
A couple of niggles, nothing major jumping out at me

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 2 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:277]: Aha, bad translation from https://github.com/ansible-collections/amazon.aws/blob/main/plugins/modules/ec2_vpc_dhcp_option.py#L251 - thanks!

File: tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml (hunk 3 above)
Comment (jillr) [reply tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml:268]: Sorry, I may just be being dense but I'm not sure what you're asking on this.  We should only have these 3 keys here because those should be the only 3 keys that are configured in this option set's DhcpConfiguration at this point.

Are you asking if we should return the `normalized_config()` (which attempts to preserve `new_config` from the original module insofar as `new_config` wasn't even entirely consistent in the first place) as well as just dumping the boto3 response?

Comment (tremble) [reply tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml:268]: > We should only have these 3 keys here because those should be the only 3 keys that are configured in this option set's DhcpConfiguration at this point.

Ah, ok so it's **supposed** to have done something approximating `purge_none_params`

Most of the time we do `- '"foo" in bar.baz'` to say we only care that the mentioned keys are there, not that it must be **only** specific keys (it gives us flexibility if Amazon add extra features).  I'm worried about non-standard examples knocking about that'll confuse people.  I'd be good with just a comment that "this is non-standard behaviour because..."

> Are you asking if we should return the `normalized_config()` (which attempts to preserve `new_config` from the original module insofar as `new_config` wasn't even entirely consistent in the first place) as well as just dumping the boto3 response?

In general, Yes.

Comment (jillr) [reply tests/integration/targets/ec2_vpc_dhcp_option/tasks/main.yml:268]: Gotcha.  Added a quick take at that, no tests yet (in the current incarnation it would solve the need to return tags at least).  I really don't like the whole Key, Values, Value structure that amazon is giving us here but that's what we have to work with.

**Comment by jillr** [comment]
@tremble Took a swipe at returning both result formats in both modules as discussed.

File: plugins/module_utils/ec2.py [hunk 4]
```diff
@@ -805,3 +805,46 @@ def compare_aws_tags(current_tags_dict, new_tags_dict, purge_tags=True):
...
+    The boto2 module returned a config dict, but boto3 returns a list of dicts
+    Make the data we return look like the old way, so we don't break users.
+    This is also much more user-friendly.
+    boto3:
+        'DhcpConfigurations': [
+            {'Key': 'domain-name', 'Values': [{'Value': 'us-west-2.compute.internal'}]},
+            {'Key': 'domain-name-servers', 'Values': [{'Value': 'AmazonProvidedDNS'}]},
+            {'Key': 'netbios-name-servers', 'Values': [{'Value': '1.2.3.4'}, {'Value': '5.6.7.8'}]},
+            {'Key': 'netbios-node-type', 'Values': [1]},
+             {'Key': 'ntp-servers', 'Values': [{'Value': '1.2.3.4'}, {'Value': '5.6.7.8'}]}
```
Comment (markuman) [inline plugins/module_utils/ec2.py:821]: one whitespace too much , but not worth an extra commit

File: plugins/module_utils/ec2.py [hunk 5]
```diff
@@ -805,3 +805,46 @@ def compare_aws_tags(current_tags_dict, new_tags_dict, purge_tags=True):
...
+        },
+    """
+    config_data = {}
+
+    if len(option_config) == 0:
+        # If there is no provided config, return the empty dictionary
+        return config_data
+
+    for config_item in option_config:
+        # # Handle single value keys
```
Comment (markuman) [inline plugins/module_utils/ec2.py:839]: one `#` too much , but not worth an extra commit

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 6]
```diff
@@ -190,91 +245,233 @@
...
-                continue
-            raise e
+    current_tags = boto3_tag_list_to_ansible_dict(client.describe_tags(aws_retry=True, Filters=[{'Name': 'resource-id', 'Values': [dhcp_options_id]}])['Tags'])
+
+    if tags:
+        tags_to_set, tags_to_unset = compare_aws_tags(current_tags, tags, purge_tags=purge_tags)
+    if purge_tags and not tags:
+        tags_to_unset = current_tags
+
+    if tags_to_unset:
```
Comment (markuman) [inline plugins/modules/ec2_vpc_dhcp_option.py:279]: as fas as I see `if tags_to_unset:` is not necessary here. The logic can instantly applied after L276 `if purge_tags and not tags`  
but on the other hand, the code logic looks cleaner and more separated.

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 7]
```diff
@@ -190,91 +245,233 @@
...
-    if len(associations) > 0:
-        return False
-    else:
-        vpc_conn.delete_dhcp_options(dhcp_options_id)
-        return True
+    new_config = []
+    params = module.params
+    if params['domain_name'] is not None:
+        new_config.append({'Key': 'domain-name', 'Values': [{'Value': params['domain_name']}]})
+    if params['dns_servers'] is not None:
```
Comment (markuman) [inline plugins/modules/ec2_vpc_dhcp_option.py:374]: what about just `params.get('domain_name'):` and reduce `if` and `for`in single `for server in params.get('dns_servers'):` 
but yeah, then the extra vars (`dns_server_list`, `ntp_server_list` etc are always be initialized, even if not used.

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 8]
```diff
@@ -190,91 +245,233 @@
...
+            netbios_server_list.append({'Value': server})
+        new_config.append({'Key': 'netbios-name-servers', 'Values': netbios_server_list})
+    if params['netbios_node_type'] is not None:
+        new_config.append({'Key': 'netbios-node-type', 'Values': params['netbios_node_type']})
+
+    return new_config
+
+
+def create_dhcp_option_set(client, module, new_config):
+    changed = True
```
Comment (markuman) [inline plugins/modules/ec2_vpc_dhcp_option.py:396]: code after comment block?

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 9]
```diff
@@ -190,91 +245,233 @@
...
+    This is the only place we use it, so create it now
+    https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateDhcpOptions.html
+    We have to do this after inheriting any existing_config, so we need to start with the object
+    that we made in create_dhcp_config().
+    normalize_config() gives us the nicest format to work with for this.
+    """
+    desired_config = normalize_ec2_vpc_dhcp_config(new_config)
+    create_config = []
+    for option in ['domain-name', 'domain-name-servers', 'ntp-servers', 'netbios-name-servers']:
+        if desired_config.get(option) is not None:
```
Comment (markuman) [inline plugins/modules/ec2_vpc_dhcp_option.py:408]: is `is not None` necessary here?

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 10]
```diff
@@ -298,116 +496,82 @@ def main():
         supports_check_mode=True
     )
 
-    params = module.params
+    vpc_id = module.params['vpc_id']
```
Comment (markuman) [inline plugins/modules/ec2_vpc_dhcp_option.py:499]: Can we drop `default=None` when we replace here everything with `module.params(paraname)` or does it break some backwards compatibility?

**Review by markuman — CHANGES_REQUESTED** [review]
In `create_dhcp_option_set()`, the code must start after the code block.

All other comments might be cosmetics.

**Comment by jillr** [comment]
I'm going to have to dig into the CI tests that are failing on netcommon imports; they shouldn't be related to anything happening here

**Review by goneri — APPROVED** [review]
Looks good to me, even if we need to delay the merge because of the breaking changes.

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 6 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:279]: My concern is keeping things separated enough for the changed logic to be clear.  I'm not sure where I would like to put the changed=True to cover a case where there are no tags_to_set but there are tags_to_unset without potentially setting the wrong changed in other cases.  It might be a bit verbose this way but it's consistent with what we do in other modules, my inclination would be to leave it a bit more easy to read.

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 7 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:374]: Historically the default value of those options is None, not an empty list, and we can't iterate on None.  If I change the default to [] I get cascading effects in the data transformations where null values get set on options that should be unset, which breaks changed status assertions.  It should be possible to make this change but it would require changing code in several places unfortunately.

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 9 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:408]: No, it's not - thanks

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 10 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:499]: Do you mean like, change the arg_spec to have no defaults for these params and instead here do something like `vpc_id = module.params.get('vpc_id', None)`?

**Comment by jillr** [comment]
Thanks very much for the reviews everyone! Hopefully we're close now.

**Review by markuman — APPROVED** [review]
LGTM

File: changelogs/fragments/252_boto3_refactor_ec2_vpc_dhcp_option.yaml [hunk 11]
```diff
@@ -0,0 +1,6 @@
+breaking_changes:
+  - ec2_vpc_dhcp_option_info - Now preserves case for tag keys in return value. (https://github.com/ansible-collections/amazon.aws/pull/252)
+  - ec2_vpc_dhcp_option - The module has been refactored to use boto3. Keys and value types returned by the module are now consistent, which is a change from the previous behaviour. A ``purge_tags`` option has been added, which defaults to ``True``.  (https://github.com/ansible-collections/amazon.aws/pull/252)
+minor_changes:
+  - ec2_vpc_dhcp_option - Now also returns a boto3-style resource description in the ``dhcp_options`` result key.  This includes any tags for the ``dhcp_options_id`` and has the same format as the current return value of ``ec2_vpc_dhcp_option_info``.
+  - ec2_vpc_dhcp_option_info - Now also returns a user-friendly ``dhcp_config`` key that matches the historical ``new_config`` key from ec2_vpc_dhcp_option, and alleviates the need to use ``items2dict(key_name='key', value_name='values')`` when parsing the output of the module.
```
Comment (tremble) [inline changelogs/fragments/252_boto3_refactor_ec2_vpc_dhcp_option.yaml:6]:
```suggestion
  - ec2_vpc_dhcp_option - Now also returns a boto3-style resource description in the ``dhcp_options`` result key.  This includes any tags for the ``dhcp_options_id`` and has the same format as the current return value of ``ec2_vpc_dhcp_option_info`` (https://github.com/ansible-collections/amazon.aws/pull/252).
  - ec2_vpc_dhcp_option_info - Now also returns a user-friendly ``dhcp_config`` key that matches the historical ``new_config`` key from ec2_vpc_dhcp_option, and alleviates the need to use ``items2dict(key_name='key', value_name='values')`` when parsing the output of the module (https://github.com/ansible-collections/amazon.aws/pull/252).
```

File: plugins/modules/ec2_vpc_dhcp_option.py [hunk 12]
```diff
@@ -104,27 +110,76 @@
...
+              - '{"Key": "Collection", "Value": "amazon.aws"}'
 dhcp_options_id:
     description: The aws resource id of the primary DCHP options set created, found or removed
     type: str
     returned: when available
-changed:
-    description: Whether the dhcp options were changed
-    type: bool
-    returned: always
+new_options:
```
Comment (tremble) [inline plugins/modules/ec2_vpc_dhcp_option.py:146]: Since ec2_vpc_dhcp_option_info returns `dhcp_options` and `dhcp_config` would it be better to rename this to dhcp_config, while returning new_options for backwards compatibility?

ie. return `dhcp_options`, `dhcp_config` and `new_options`, but only document dhcp_options and dhcp_config In theory we could then deprecate new_options with something like release 3.x, maybe with a parameter to choose to return it or not.

**Review by tremble — CHANGES_REQUESTED** [review]
Minor tweaks to changelog and thoughts about what's returned.

File: plugins/modules/ec2_vpc_dhcp_option.py (hunk 12 above)
Comment (jillr) [reply plugins/modules/ec2_vpc_dhcp_option.py:146]: Yeah that makes sense

**Review by tremble — APPROVED** [review]

File: plugins/modules/ec2_vpc_dhcp_option_info.py [hunk 13]
```diff
@@ -69,10 +69,69 @@
...
+         owner_id:
+             description: The ID of the AWS account that owns the DHCP options set.
+             type: str
+             sample: 012345678912
+         tags:
+             description: The tags to be applied to a DHCP options set
+             type: list
+             sample:
+               - '{"Key": "CreatedBy", "Value": "ansible-test"}'
+               - '{"Key": "Collection", "Value": "amazon.aws"}'
```
Comment (tremble) [inline plugins/modules/ec2_vpc_dhcp_option_info.py:96]:
```suggestion
    description: The DHCP options created, associated or found
    returned: always
    type: list
    elements: dict
    contains:
        dhcp_configurations:
            description: The DHCP configuration for the option set
            type: list
            sample:
              - '{"key": "ntp-servers", "values": [{"value": "10.0.0.2" , "value": "10.0.1.2"}]}'
              - '{"key": "netbios-name-servers", "values": [{value": "10.0.0.1"}, {"value": "10.0.1.1" }]}'
        dhcp_options_id:
            description: The aws resource id of the primary DCHP options set created or found
            type: str
            sample: "dopt-0955331de6a20dd07"
        owner_id:
            description: The ID of the AWS account that owns the DHCP options set.
            type: str
            sample: 012345678912
        tags:
            description: The tags to be applied to a DHCP options set
            type: list
            sample:
              - '{"Key": "CreatedBy", "Value": "ansible-test"}'
              - '{"Key": "Collection", "Value": "amazon.aws"}'
```

**Review by tremble — COMMENTED** [review]
whitespace fixup

File: plugins/modules/ec2_vpc_dhcp_option_info.py [hunk 14]
```diff
@@ -69,10 +69,69 @@
...
+            description: The tags to be applied to a DHCP options set
+            type: list
+            sample:
+              - '{"Key": "CreatedBy", "Value": "ansible-test"}'
+              - '{"Key": "Collection", "Value": "amazon.aws"}'
+dhcp_config:
+    description: The boto2-style DHCP options created, associated or found. Provided for consistency with ec2_vpc_dhcp_option's `new_config`.
+    returned: always
+    type: list
+    contains: dict
```
Comment (tremble) [inline plugins/modules/ec2_vpc_dhcp_option_info.py:101]:
```suggestion
```

**Comment by jillr** [comment]
rebased around the git conflict.
